crates/type/src/digest/digests.rs

```diff
@@ -0,0 +1,186 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: Apache-2.0
+
+use super::{Algorithm, Error, Reader, Verifier, Writer};
+
+use std::collections::btree_map::IntoIter;
+use std::collections::BTreeMap;
+use std::ops::{Deref, DerefMut};
+use std::str::FromStr;
+
+use drawbridge_byte::Bytes;
+use serde::{Deserialize, Serialize};
+
+#[cfg(all(feature = "headers", feature = "http"))]
+use headers::{Error as HeadErr, Header, HeaderName, HeaderValue};
+
+/// A set of hashes for the same contents
+#[derive(Clone, Default, Debug, Serialize, Deserialize)]
+pub struct ContentDigest<H = Box<[u8]>>(BTreeMap<Algorithm, Bytes<H>>)
+where
+    H: AsRef<[u8]> + From<Vec<u8>>;
+
+impl<H> ContentDigest<H>
+where
+    H: AsRef<[u8]> + From<Vec<u8>>,
+{
+    /// Creates a reader instance
+    pub fn reader<T>(&self, reader: T) -> Reader<T> {
+        Reader::new(reader, self.iter().map(|x| *x.0))
+    }
+
+    /// Creates a writer instance
+    pub fn writer<T>(&self, writer: T) -> Writer<T> {
+        Writer::new(writer, self.iter().map(|x| *x.0))
+    }
+
+    /// Creates a verifier instance
+    pub fn verifier<T>(self, reader: T) -> Verifier<T, H> {
+        Verifier::new(self.reader(reader), self)
+    }
+}
+
+impl<H> From<BTreeMap<Algorithm, Bytes<H>>> for ContentDigest<H>
+where
+    H: AsRef<[u8]> + From<Vec<u8>>,
+{
+    fn from(value: BTreeMap<Algorithm, Bytes<H>>) -> Self {
+        Self(value)
+    }
+}
+
+impl<H> Eq for ContentDigest<H> where H: AsRef<[u8]> + From<Vec<u8>> {}
+impl<T, U> PartialEq<ContentDigest<U>> for ContentDigest<T>
+where
+    T: AsRef<[u8]> + From<Vec<u8>>,
+    U: AsRef<[u8]> + From<Vec<u8>>,
+{
+    fn eq(&self, other: &ContentDigest<U>) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+
+        for (lhs, rhs) in self.0.iter().zip(other.0.iter()) {
+            if lhs.0 != rhs.0 {
+                return false;
+            }
+
+            if lhs.1.as_ref() != rhs.1.as_ref() {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl<H> Deref for ContentDigest<H>
+where
+    H: AsRef<[u8]> + From<Vec<u8>>,
+{
+    type Target = BTreeMap<Algorithm, Bytes<H>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<H> DerefMut for ContentDigest<H>
+where
+    H: AsRef<[u8]> + From<Vec<u8>>,
+{
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<H> std::fmt::Display for ContentDigest<H>
+where
+    H: AsRef<[u8]> + From<Vec<u8>>,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut comma = "";
+
+        for (algo, hash) in self.iter() {
+            write!(f, "{}{}=:{}:", comma, algo, hash)?;
+            comma = ",";
+        }
+
+        Ok(())
```

Comment:
Do we want to separate with `\,` as in the RFC draft? 

Also, a style nit: `comma` variable here is a bit confusing
We could also do something like this instead:
```diff
diff --git a/crates/type/src/digest/digests.rs b/crates/type/src/digest/digests.rs
index 4a40f15..0534a0e 100644
--- a/crates/type/src/digest/digests.rs
+++ b/crates/type/src/digest/digests.rs
@@ -99,13 +99,15 @@ where
     H: AsRef<[u8]> + From<Vec<u8>>,
 {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        let mut comma = "";
-
-        for (algo, hash) in self.iter() {
-            write!(f, "{}{}=:{}:", comma, algo, hash)?;
-            comma = ",";
+        let mut iter = self
+            .iter()
+            .map(|(algo, hash)| format!("{}=:{}:", algo, hash));
+        if let Some(s) = iter.next() {
+            write!(f, "{}", s)?;
+        }
+        for s in iter {
+            write!(f, ",{}", s)?;
         }
-
         Ok(())
     }
 }
```
What do you think?

Reply:
The iterator approach forces unnecessary allocations. I'm happy with `sep`. I'm honestly not sure what is correct, `\,` or `,`. Please research and tell me what to do.

Reply:
@rvolosatovs Can we merge this and you can fix in a follow up?